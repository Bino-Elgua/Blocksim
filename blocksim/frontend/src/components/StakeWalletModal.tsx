import React, { useState } from "react";
import axios from "axios";
import styles from "./StakeWalletModal.module.css";

interface Props {
  onStake: (walletId: string, amount: number) => void;
}

export const StakeWalletModal: React.FC<Props> = ({ onStake }) => {
  const [walletId, setWalletId] = useState<string>("");
  const [stakeAmount, setStakeAmount] = useState<number>(10);
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);

  const handleStake = async () => {
    if (!walletId.trim()) {
      alert("Enter wallet ID");
      return;
    }
    setLoading(true);
    try {
      const res = await axios.post("/api/chain/stake", { wallet_id: walletId, amount: stakeAmount });
      if (res.data?.status === "staked") {
        onStake(walletId, stakeAmount);
        setIsOpen(false);
        alert("Staked successfully!");
      } else {
        alert("Unexpected response: " + JSON.stringify(res.data));
      }
    } catch (err: any) {
      alert("Stake failed: " + (err?.message ?? JSON.stringify(err)));
    } finally {
      setLoading(false);
    }
  };

  return (
    <div>
      <button onClick={() => setIsOpen(true)}>Stake to Race</button>
      {isOpen && (
        <div className={styles.modalOverlay} onClick={() => setIsOpen(false)}>
          <div className={styles.modalContent} onClick={(e: React.MouseEvent) => e.stopPropagation()}>
            <h3>Stake Wallet</h3>
            <input value={walletId} onChange={(e) => setWalletId(e.target.value)} placeholder="Wallet ID (e.g., 0x123...)" />
            <input value={stakeAmount} onChange={(e) => setStakeAmount(Number(e.target.value) || 10)} type="number" placeholder="Amount (tokens)" />
            <div className={styles.actions}>
              <button onClick={handleStake} disabled={loading}>{loading ? "Staking..." : "Confirm"}</button>
              <button onClick={() => setIsOpen(false)}>Cancel</button>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};
