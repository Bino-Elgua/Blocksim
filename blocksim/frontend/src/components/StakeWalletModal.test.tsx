import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import axios from 'axios';
import { StakeWalletModal } from './StakeWalletModal';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('StakeWalletModal', () => {
  test('opens modal and submits stake successfully', async () => {
    mockedAxios.post.mockResolvedValueOnce({ data: { status: 'staked' } });

    const onStake = jest.fn();
    render(<StakeWalletModal onStake={onStake} />);

    // open modal
    fireEvent.click(screen.getByText('Stake to Race'));
    expect(screen.getByPlaceholderText(/Wallet ID/)).toBeInTheDocument();

    fireEvent.change(screen.getByPlaceholderText(/Wallet ID/), { target: { value: '0xabc' } });
    fireEvent.change(screen.getByPlaceholderText(/Amount/), { target: { value: '42' } });

    fireEvent.click(screen.getByText('Confirm'));

    await waitFor(() => expect(onStake).toHaveBeenCalledWith('0xabc', 42));
  });

  test('shows toast on error response', async () => {
    mockedAxios.post.mockResolvedValueOnce({ data: { status: 'failed', reason: 'insufficient' } });
    const onStake = jest.fn();
    const { container } = render(<StakeWalletModal onStake={onStake} />);
    fireEvent.click(screen.getByText('Stake to Race'));
    fireEvent.change(screen.getByPlaceholderText(/Wallet ID/), { target: { value: '0xdef' } });
    fireEvent.click(screen.getByText('Confirm'));
    await waitFor(() => expect(screen.getByText(/Stake failed|Unexpected response/i)).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });
});
