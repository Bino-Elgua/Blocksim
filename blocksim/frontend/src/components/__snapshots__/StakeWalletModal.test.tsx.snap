// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakeWalletModal shows toast on error response 1`] = `
<div>
  <div>
    <div
      style="position: fixed; top: 16px; right: 16px; background: rgb(15, 23, 42); color: rgb(226, 232, 240); padding: 8px 12px; border-radius: 6px; box-shadow: 0 4px 12px rgba(0,0,0,0.2);"
    >
      <div>
        Unexpected response: {"status":"failed","reason":"insufficient"}
      </div>
      <button
        style="margin-left: 8px;"
      >
        Close
      </button>
    </div>
    <button>
      Stake to Race
    </button>
    <div
      class="modalOverlay"
    >
      <div
        class="modalContent"
      >
        <h3>
          Stake Wallet
        </h3>
        <input
          placeholder="Wallet ID (e.g., 0x123...)"
          value="0xdef"
        />
        <input
          placeholder="Amount (tokens)"
          type="number"
          value="10"
        />
        <div
          class="actions"
        >
          <button>
            Confirm
          </button>
          <button>
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
